### ISSUE TITLE
auto generated alias converts some special characters to numbers

### ISSUE THREAD

**Issue by fritzmg** [contao/core-bundle#1334]
__Reproduction:__

1. Log into the Contao Online Demo.
2. Create a new form in the form generator.
3. Set the alias to `Newsletter Anmeldung (de)`.
4. Click save.

The auto generated alias will be `newsletter-anmeldung-40-de-41`.

_Note:_ only some characters will actually be converted. For instance, the alias `!"§$%&/()=?` will be converted to `40-41-61`.

**Comment by ausi** [comment]
Could you please check if replacing Line https://github.com/contao/core-bundle/blob/2b9d7e5ce1da2cb78d01eaa2356fb32a983053f5/src/Resources/contao/dca/tl_form.php#L464 with the following code solves the issue?

```php
$varValue = System::getContainer()->get('contao.slug.generator')->generate(html_entity_decode(StringUtil::stripInsertTags($dc->activeRecord->title)), $slugOptions);
```

**Comment by fritzmg** [comment]
Yes, that solves the issue.

**Comment by fritzmg** [comment]
However, shouldn't something like `StringUtil::generateAlias` be reintroduced, so that you don't have to do
```php
html_entity_decode(StringUtil::stripInsertTags(…))
```
in every DCA?

The old `StringUtil::generateAlias` function used to do:

1. `StringUtil::decodeEntities`
2. `StringUtil::restoreBasicEntities`
3. `standardize`
    1. `html_entity_decode`
    2. `strip_insert_tags`
    3. `utf8_romanize`
    4. `preg_replace`

**Comment by ausi** [comment]
The correct order should be:

1. strip insert tags
2. restore basic entities
3. decode entities
4. generate slug

As suggested in https://github.com/contao/core-bundle/pull/1016#issuecomment-359384997 should we an optional parameter to `StringUtil::generateAlias()` for the PageModel? So that we only have to call `StringUtil::generateAlias($dc->activeRecord->title, $objPage)`

**Comment by m-vo** [comment]
Can't we put that into a service? This would imho make it easier for extensions to adapt.

**Comment by ausi** [comment]
How should it be called `contao.slug.helper`, `contao.slug.alias_generator`,...?

**Comment by fritzmg** [comment]
Well, why not `contao.slug.generator`? I know that service name is currently `Ausi\SlugGenerator\SlugGenerator`, but it could be something like `Contao\CoreBundle\Slug\Generator` instead, which in turn uses `Ausi\SlugGenerator\SlugGenerator`. Seems superfluous to register two services for that when one won't be used anymore (will it?)

**Comment by backbone87** [comment]
if we introduce a contao specific class we should stick to contao terminology: AliasGenerator (which internally uses @ausi's slug generator)

**Comment by ausi** [comment]
We cannot remove the `contao.slug.generator` service, this would break backwards compatibility.

**Comment by aschempp** [comment]
> We cannot remove the contao.slug.generator service, this would break backwards compatibility.

Can we not decorate it?

**Comment by ausi** [comment]
We can. We could also replace the service with our own class that extends `Ausi\SlugGenerator\SlugGenerator`.

**Comment by backbone87** [comment]
is there an interface for your SlugGenerator?

**Comment by ausi** [comment]
No.

**Comment by leofeyer** [comment]
As discussed in the developer's meeting, this is what we want:

```php
// tl_form_field.php
public function generateAlias($varValue, DataContainer $dc)
{
	$autoAlias = false;

	// Generate an alias if there is none
	if ($varValue == '')
	{
		$autoAlias = true;
		$slugOptions = array();

		// Read the slug options from the associated page
		if (($objPage = PageModel::findWithDetails($dc->activeRecord->jumpTo)) !== null)
		{
			$slugOptions = $objPage->getSlugOptions();
		}

		$varValue = System::getContainer()->get('contao.slug.generator')->generate(StringUtil::prepareSlug($dc->activeRecord->title), $slugOptions);
	}
```

**Comment by leofeyer** [comment]
Implemented in 9d7bdf494e929b97cd7992c22901a440c7ad34e6.
